### ISSUE TITLE
Group Query Attention

### ISSUE THREAD

**Issue by Zaire404** [Zaire404/docs#17]
<img width="1080" height="336" alt="Image" src="https://github.com/user-attachments/assets/79894aca-3113-46dc-9dd7-163c2b1bf888" />

<img width="5672" height="1754" alt="Image" src="https://github.com/user-attachments/assets/ab883209-a4d4-4630-bf47-970a828d0233" />

### 1. 设定场景参数

为了方便心算和理解，我们不使用 Llama-3 那种巨大的参数（4096维），而是设定一个**迷你模型**：

* **模型隐藏层维度 (`hidden_size`)**: **512**
* **Query 头数 (`num_attention_heads`, )**: **8** 个头
* **KV 头数 (`num_key_value_heads`, )**: **2** 个头
* **单头维度 (`head_dim`, )**:  **64** 维

> **这就是 GQA 的核心特征**：Query 有 8 个头，但 K 和 V 只有 2 个头。这意味着 **4 个 Query 头要共享 1 个 KV 头**，我们称之为 **Group Size = 4**。

---

### 2. 一个 query_head 处理多少维？

**答案是：64 维。**

不管模型有多大，**一个注意力头（Attention Head）内部处理的数据维度永远是 `head_dim`**。
虽然输入向量是 512 维的，但在进入注意力计算之前，它会被切分（或者说投影）成 8 个 64 维的向量，分给 8 个 Query 头。

---

### 3. GQA 的详细计算流程（一步步举例）

假设输入是一个 Token 的向量 ，形状是 `[1, 512]`（Batch=1, Seq=1, Dim=512）。

#### 第一步：线性投影 (Linear Projection)

输入 $X$  分别乘以  $W_q$, $W_k$, $W_v$ 权重矩阵。

* **生成 Q**： $X \times W_q \rightarrow$ 得到 **8 个 ** Q 向量，每个 **64 维**。
* 记为： $Q_0, Q_1, Q_2, Q_3, Q_4, Q_5, Q_6, Q_7$

* **生成 K**： $X \times W_k \rightarrow$ 得到 **2 个** K 向量，每个 **64 维**。
* 记为： $K_A, K_B$

* **生成 V**： $X \times W_v \rightarrow$  得到 **2 个** V 向量，每个 **64 维**。
* 记为： $V_A, V_B$

#### 第二步：分组 (Grouping / Broadcast)

这是 GQA 的关键步骤。系统需要把 8 个 Q 分配给 2 个 KV 组。

* **第 1 组 (Group A)**：
* 包含 Query 头： $Q_0, Q_1, Q_2, Q_3$
* **共享** Key/Value： $K_A, V_A$

* **第 2 组 (Group B)**：
* 包含 Query 头： $Q_4, Q_5, Q_6, Q_7$
* **共享** Key/Value： $K_B, V_B$

在实际计算（如 PyTorch/vLLM）中，这一步通常通过**逻辑上的“复制”**来实现。虽然物理显存中只有 1 个 ，但计算时 $Q_0$ 到 $Q_3$ 都会去读这同一个 。

#### 第三步：计算注意力分数 (Attention Scores)

现在开始算“点积”（Dot Product）。每个 Q 都要找它的对象 K 做乘法。

* **对于第 1 组：**
*  $Score_0 = \text{Softmax}(Q_0 \cdot K_A^T)$
*  $Score_1 = \text{Softmax}(Q_1 \cdot K_A^T)$ <-- 注意：用的还是 $K_A$
*  $Score_2 = \text{Softmax}(Q_2 \cdot K_A^T)$ <-- 注意：用的还是 $K_A$
*  $Score_3 = \text{Softmax}(Q_3 \cdot K_A^T)$ <-- 注意：用的还是 $K_A$

* **对于第 2 组：**
* $Score_4 = \text{Softmax}(Q_4 \cdot K_B^T)$ 
* ...
* $Score_7 = \text{Softmax}(Q_7 \cdot K_B^T)$

每个 Score 都是一个标量（或者在长序列下是一个权重向量）。

#### 第四步：加权求和得到输出 (Output)

用算出来的 Score 去乘 Value。

* **第 1 组输出：**
* $Out_0 = Score_0 \times V_A$  <-- 结果是一个 64 维向量
* $Out_1 = Score_1 \times V_A$
* ...

* **第 2 组输出：**
* $Out_4 = Score_4 \times V_B$
* ...

虽然 $Q_0$ 和 $Q_1$ 看了同一个 $K_A$ 和 $V_A$ ，但因为 $Q$ 向量本身不同（ $Q_0 \neq Q_1$ ），所以算出来的 $Score$ 不同，最终得到的 $Out_0$ 和 $Out_1$ 也是不同的。这就保留了“多头”的多样性。

#### 第五步：拼接 (Concat)

现在我们手里有 8 个输出向量 ( $Out_0 \dots Out_7$ )，每个都是 64 维。
我们将它们拼回去：

$$8 \times 64 = 512$$

最终恢复成 **512 维** 的向量，准备传给下一层。

### 总结图示

```text
输入 (512维)
      |
[投影层]
      |
-----------------------------
|         |        |        |
Q (8个)   K (2个)  V (2个)  <--- 只有Q是全量的
|         |        |
|--Q0--\  |        |
|--Q1---| K_A      | V_A    <--- 第1组：4个Q共用一套KV
|--Q2---|          |
|--Q3--/           |
                   |
|--Q4--\           |
|--Q5---| K_B      | V_B    <--- 第2组：4个Q共用一套KV
|--Q6---|          |
|--Q7--/           |
-----------------------------
      |
[独立计算 Attention]
      |
输出 8 个 64维向量
      |
[拼接] -> 恢复 512维

```

### 为什么这么做？

在这个例子中，我们要存的 KV Cache（缓存）大小从 **8 份减少到了 2 份**。显存占用直接变成了原来的 **1/4**，读取速度也快了 4 倍，但因为有 8 个不同的 Q 在提问，模型的智商（理解能力）并没有下降太多。
